### ISSUE TITLE
What is the correct way to calculate validation loss?

### ISSUE THREAD

**Issue by vedantroy** [lucidrains/imagen-pytorch#115]
Hi! 
I was wondering what is the correct way to calculate validation loss when using the `ImagenTrainer` without using the `valid_step` method. I'm having trouble using that method because I'm not sure what the format of the returned data from the validation data loader should be. (I normally don't use a training dataloader)

I tried doing:

```
with torch.no_grad():
     # call trainer here
```

but that doesn't make sense since it looks like the trainer updates the weights inside of its `forward` method.
I'm thinking of trying:

```
with torch.no_grad():
     validation_loss = trainer.imagen(images=foo, texts=bar)
```

Although that seems like it might be missing some of the `ema` things.

Side note, thanks for answering all of my questions. It's been super helpful :)

**Comment by lucidrains** [comment]
@vedantroy Hi Vedant :) You are welcome!

So the way to do validation loss is to have a separate dataloader, identical to training dataloader, but with data that the model will not have seen during training. You can also just split off a chunk of the data from your existing training dataloader using this function https://pytorch.org/docs/stable/data.html#torch.utils.data.random_split (anywhere from 2-5% is good enough)

As long as you set your trainer to eval `trainer.eval()` before calculating the validation loss, it will not update `https://github.com/lucidrains/imagen-pytorch/blob/main/imagen_pytorch/trainer.py#L730`

**Comment by lucidrains** [comment]
@vedantroy yea that's fine, for your validation phase, you can do exactly the same thing (just average your loss across batches)

**Comment by vedantroy** [comment]
> @vedantroy yea that's fine, for your validation phase, you can do exactly the same thing (just average your loss across batches)

Got it; and presumably I just run `trainer.eval()` and `trainer.train() before/after the validation phase?

**Comment by lucidrains** [comment]
@vedantroy yup, correct!
